src/shogun/mathematics/linalg/internal/implementation/RangeFill.h

```diff
@@ -0,0 +1,134 @@
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ *
+ * The views and conclusions contained in the software and documentation are those
+ * of the authors and should not be interpreted as representing official policies,
+ * either expressed or implied, of the Shogun Development Team.
+ */
+
+#ifndef RANGE_FILL_IMPL_H_
+#define RANGE_FILL_IMPL_H_
+
+#include <shogun/lib/config.h>
+#include <shogun/lib/SGVector.h>
+#include <shogun/lib/SGMatrix.h>
+
+#include <numeric>
+
+namespace shogun
+{
+
+namespace linalg
+{
+
+namespace implementation
+{
+
+/**
+ * @brief Generic class which is specialized for different backends to perform addition
+ */
+template <enum Backend, class Matrix>
+struct range_fill
+{
+	/**Scalar type */
+	typedef typename Matrix::Scalar T;
+
+	/**Range fill a vector or a matrix with start...start+len-1
+     * @param A - the matrix to be filled
+     * @param len - length of the matrix to be filled
+	 * @param start - value to be assigned to first element of vector or matrix
+	 */	
+	static void compute(Matrix A, T start);
+
+};
+
+/**
+ *@brief Partial specialization of add for the Native backend
+ */
+template <class Matrix>
+struct range_fill<Backend::NATIVE, Matrix>
+{
+	/** Scalar type */
+	typedef typename Matrix::Scalar T;
+
+	/** Range fill a matrix with start...start+len-1
+     * @param A - the matrix to be filled
+     * @param len - length of the matrix to be filled
+	 * @param start - value to be assigned to first element of vector or matrix
+	 */	
+	static void compute(SGMatrix<T> A, T start)
+	{
+		compute(A.matrix, A.num_rows*A.num_cols, start);
+	}
+
+	/** Range fill a vector with start...start+len-1
+     * @param A - the matrix to be filled
+     * @param len - length of the matrix to be filled
+	 * @param start - value to be assigned to first element of vector or matrix
+	 */	
+	static void compute(SGVector<T> A, T start)
+	{
+		compute(A.vector, A.vlen, start);
+	}
+
+	/**Range fill a vector or a matrix with start...start+len-1
+     * @param A - the matrix to be filled
+     * @param len - length of the matrix to be filled
+	 * @param start - value to be assigned to first element of vector or matrix
+	 */	
+	static void compute(T* A, index_t len, T start)
+	{
+		std::iota(A, A+len, start);
+	}
+
+};
+
+/**
+ * @brief Generic class which is specialized for different backends to perform addition
+ */
+template <enum Backend, class Vector>
+struct range_fill_vec
+{
```

Comment:
Okay here is the plan. 

- I think using separate methods for vectors and matrix is a not a very good idea, leads to a lot of code duplication. Their underlying data is just a plain memory array. Also, in shogun, matrices are always column-major. So treating them in the same manner shouldn't be an issue. I admit that it is done in linalg in quite a few places but that's bad. We should change that.  I just added methods `T* data()` and `index_t size()` in both SGMatrix and SGVector (check PR #2830). So using these we can handle the underlying memory exactly the same way as one would do with raw array types (once that PR is merged).

- Write one struct 
```c++
template <enum Backend, class Container>
struct range_fill
{
    // you don't need to put anything here. Your code should go only with
    // the specializations.
};
```
and partially specialize NATIVE. Access the data using `container.data()` and size using `container.size()`.

- Write two wrappers
```c++
template <Backend backend, class Container> 
range_fill(Container container, typename Container::Scalar begin=0)
{
}
```
and
```c++
template <typename T> // backend not required. raw pointers should work only with NATIVE
range_fill(T* memarray, index_t size, T begin=0)
{
    // construct a SGVector object with refcounting turned off
    SGVector<T> v(memarray, size, false);
    // instantiate the impl template with SGVector<T> type and NATIVE backend and call compute
}
```
- Since this supports NATIVE implementation, keep these above methods outside of the `HAVE_LINALG_LIB` guard (since they can be invoked even in absence of external linear algebra lib).